src/qtcwin.c

```diff
@@ -0,0 +1,1816 @@
+        if (msgidx > 2 && msgidx < 6) {
+	    msgidx = 3;
+	}
+	mvwprintw(qtcwin, ++j, 36, help_send_msgs[msgidx]);
+    }
+    wattrset(qtcwin, line_inverted);
+    mvwprintw(qtcwin, ++j, 36, "PgUP/PgDW: QRQ/QRS");
+    if (*qtccurrdiretion == RECV) {
+	mvwprintw(qtcwin, ++j, 36, "ENTER: R & next OR AGN");
+    }
+    if (*qtccurrdiretion == SEND) {
+	mvwprintw(qtcwin, ++j, 36, "ENTER: send QTC");
+    }
+    for(i=0; i<12 && j < 12; i++) {
+	if (*qtccurrdiretion == RECV) {
+	    if (strlen(qtc_recv_msgs[i]) > 0) {
+		strncpy(buff, qtc_recv_msgs[i], strlen(qtc_recv_msgs[i])-1);
+		buff[strlen(qtc_recv_msgs[i])-1] = '\0';
+		mvwprintw(qtcwin, ++j, 36, "F%-2d: %s", (i+1), buff);
+	    }
+	}
+	if (*qtccurrdiretion == SEND) {
+	    if (strlen(qtc_send_msgs[i]) > 0) {
+		strncpy(buff, qtc_send_msgs[i], strlen(qtc_send_msgs[i])-1);
+		buff[strlen(qtc_send_msgs[i])-1] = '\0';
+		mvwprintw(qtcwin, ++j, 36, "F%-2d: %s", (i+1), buff);
+	    }
+	}
+    }
+    if (trxmode == DIGIMODE && *qtccurrdiretion == RECV) {
+        mvwprintw(qtcwin, ++j, 36, "CTRL-T: Terminal window");
+	mvwprintw(qtcwin, ++j, 36, "CTRL-S: Start capture");
+	mvwprintw(qtcwin, ++j, 36, "CTRL-E: End capture");
+    }
+
+    return 0;
+}
+
+int parse_ry_line(char * line) {
+    int j, i = 0, t = -1, f;
+    char lline[40], *token, *wtoken, ttoken[50], tokens[5][15] = {"", "", "", "", ""};	// 5 members, cause possible callsign will AB/CD1EF/P, which will use 3
+    char *saveptr1, *saveptr2;
+    char sep[6] = "";
+    int tactivefield;
+    static int nr_parsed_line = 0;
+    int serial_nr = 0;
+    int maxlength[] = {4, 14, 4};
+
+    if (*qtccount == 0 || qtcreclist.serial == 0) {
+	nr_parsed_line = 0;
+    }
+    if ((*qtccount > 0 || qtcreclist.serial > 0) && nr_parsed_line == 0) {
+	nr_parsed_line = 1;
+    }
+
+    strncpy(lline, line, strlen(line));
+    tactivefield = activefield;
+    if (nr_parsed_line == 0) {		// 1st line: SERIAL/NR
+	wtoken = strtok_r(lline, " ", &saveptr1);
+	while(wtoken != NULL && nr_parsed_line == 0) {
+	    serial_nr = 0;
+	    strcpy(ttoken, wtoken);
+	    sep[0] = '/';
+	    sep[1] = '\0';
+	    token = strtok_r(ttoken, sep, &saveptr2);
+	    t = -1;
+	    while(token != NULL && t < 2) 
+	    {
+		t++;
+		strcpy(tokens[t], token);
+		token = strtok_r(NULL, sep, &saveptr2);
+	    }
+	    f = 0;
+	    for(i=0; i<=t; i++) {
+		f = 1;
+		for(j=0; j<strlen(tokens[i]); j++) {
+		    if (! isdigit(tokens[i][j])) {
+			f = 0;
+		    }
+		}
+		if (f == 1) {
+		    tactivefield = activefield;
+		    activefield = i+1;
+		    showfield(tactivefield);
+		    for(j=0;j<strlen(tokens[i]); j++) {
+			modify_field(tokens[i][j]);
+		    }
+		    showfield(activefield);
+		    serial_nr++;
+		}
+	    }
+	    if (serial_nr == 2) {
+		nr_parsed_line++;
+	    }
+	    wtoken = strtok_r(NULL, " ", &saveptr1);
+	}
+    }
+    else {
+	strcpy(sep, "/;-: ");
+	token = strtok(lline, sep);
+	while(token != NULL && t < 5) 
+	{
+	    t++;
+	    strcpy(tokens[t], token);
```

Comment:
Tokens are generated by strtok(). If the strtok() ends the loop, then the token will be NULL, and "t < 5" will not be evaulate.
This "t < 5" os only if the RTTY line contains some extra separator character. The lenght of tokens is 5:
1422.    ..., tokens[5][15] = {"", "", "", "", ""};